File: FrameworkExample/Objective-C/VponSdkSampleObjC/VponSdkSampleObjC/Sdk/VponSdkNativeViewController.m
//
//  VponSdkNativeViewController.m
//  VponSdkSampleObjC
//
//  Created by EricChien on 2020/1/20.
//  Copyright © 2020 Soul. All rights reserved.
//

#import "VponSdkNativeViewController.h"
#import <VpadnSDKAdKit/VpadnSDKAdKit.h>
#import <AdSupport/AdSupport.h>

@interface VponSdkNativeViewController () <VponNativeAdLoaderDelegate, VponNativeAdDelegate, VponVideoControllerDelegate>

@property (weak, nonatomic) IBOutlet UIView *adContainerView;
@property (strong, nonatomic) VponNativeAdLoader *adLoader;
@property (weak, nonatomic) IBOutlet UIButton *requestButton;

@property(nonatomic, strong) VponNativeAdView *nativeAdView;

@end

@implementation VponSdkNativeViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    self.title = @"SDK - Native";
    
    _nativeAdView = [[NSBundle mainBundle] loadNibNamed:@"NativeAdView" owner:nil options:nil]
        .firstObject;
    [_adContainerView addSubview:_nativeAdView];
    _nativeAdView.translatesAutoresizingMaskIntoConstraints = NO;
    
    [NSLayoutConstraint activateConstraints:@[
        [_nativeAdView.heightAnchor constraintEqualToAnchor: _adContainerView.heightAnchor],
        [_nativeAdView.widthAnchor constraintEqualToAnchor: _adContainerView.widthAnchor]
    ]];
    
    [self requestButtonDidTouch:_requestButton];
}

// MARK: - Create VponAdRequest

- (VponAdRequest *) createRequest {
    VponAdRequest *request = [[VponAdRequest alloc] init];
    [request setUserInfoGender:VponUserGenderMale]; // 性別
    [request setUserInfoBirthdayWithYear:2000 month:8 day:17]; // 生日
    [request setContentUrl:@"https://www.vpon.com.tw/"]; // 內容
    [request setContentData:@{@"key1": @(1), @"key2": @(YES), @"key3": @"name", @"key4": @(123.31)}]; // 內容鍵值
   
    VponAdRequestConfiguration *config = VponAdRequestConfiguration.shared;
    [config setTestDeviceIdentifiers:@[[ASIdentifierManager sharedManager].advertisingIdentifier.UUIDString]]; // 取得測試廣告
    [config setTagForUnderAgeOfConsent:VponTagForUnderAgeOfConsentNotForUnderAgeOfConsent]; // 是否專為特定年齡投放
    [config setTagForChildDirectedTreatment:VponTagForChildDirectedTreatmentNotForChildDirectedTreatment]; // 是否專為兒童投放
    [config setMaxAdContentRating:VponMaxAdContentRatingGeneral]; // 最高可投放的年齡(分類)限制
  
    return request;
}

// MARK: - Button Method

- (IBAction)requestButtonDidTouch:(UIButton *)sender {
    sender.enabled = NO;
    
    _adLoader = [[VponNativeAdLoader alloc]initWithLicenseKey:@"" rootViewController:self];
    _adLoader.delegate = self;
    [_adLoader load:[self createRequest]];
}

// MARK: - VponNativeAdLoaderDelegate

- (void)adLoader:(VponNativeAdLoader *)adLoader didReceive:(VponNativeAd *)nativeAd {
    NSLog(@"adLoader didReceive nativeAd");
    _requestButton.enabled = YES;
    
    nativeAd.delegate = self;
    
    ((UILabel *)_nativeAdView.headlineView).text = nativeAd.headline;
    _nativeAdView.mediaView.mediaContent = nativeAd.mediaContent;
    
    if (nativeAd.mediaContent.hasVideoContent) {
        nativeAd.mediaContent.videoController.delegate = self;
    }
    
    ((UILabel *)_nativeAdView.bodyView).text = nativeAd.body;
    [((UIButton *)_nativeAdView.callToActionView) setTitle:nativeAd.callToAction
                                                  forState:UIControlStateNormal];
    ((UIImageView *)_nativeAdView.iconView).image = nativeAd.icon.image;
    
    // Necessary to show media content and make it clickable!
    _nativeAdView.nativeAd = nativeAd;
}

- (void)adLoader:(VponNativeAdLoader *)adLoader didFailToReceiveAdWithError:(NSError *)error {
    NSLog(@"adLoader didFailToReceiveAdWithError: %@", error.localizedDescription);
    _requestButton.enabled = YES;
}

// MARK: - VponNativeAdDelegate

- (void)nativeAdDidRecordImpression:(VponNativeAd *)nativeAd {
    NSLog(@"nativeAdDidRecordImpression");
}

- (void)nativeAdDidRecordClick:(VponNativeAd *)nativeAd {
    NSLog(@"nativeAdDidRecordClick");
}

// MARK: - VponVideoControllerDelegate

- (void)videoControllerDidPlayVideo:(VponVideoController *)videoController {
    NSLog(@"videoControllerDidPlayVideo");
}

- (void)videoControllerDidPauseVideo:(VponVideoController *)videoController {
    NSLog(@"videoControllerDidPauseVideo");
}

- (void)videoControllerDidMuteVideo:(VponVideoController *)videoController {
    NSLog(@"videoControllerDidMuteVideo");
}

- (void)videoControllerDidUnmuteVideo:(VponVideoController *)videoController {
    NSLog(@"videoControllerDidUnmuteVideo");
}

- (void)videoControllerDidEndVideoPlayback:(VponVideoController *)videoController {
    NSLog(@"videoControllerDidEndVideoPlayback");
}

@end

File: FrameworkExample/Swift/VponSdkSampleSwift/VponSdkSampleSwift/Sdk/VponSdkNativeViewController.swift
//
//  VponSdkNativeViewController.swift
//  VponSdkSampleSwift
//
//  Created by Yi-Hsiang, Chien on 2020/2/5.
//  Copyright © 2020 EricChien. All rights reserved.
//

import UIKit
import VpadnSDKAdKit
import AdSupport

class VponSdkNativeViewController: UIViewController {
    
    var adLoader: VponNativeAdLoader? // Must keep a strong reference to VponNativeAdLoader during the ad loading process!
    
    @IBOutlet weak var adContainer: UIView!
    @IBOutlet weak var requestButton: UIButton!
    
    var nativeAdView: VponNativeAdView!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        self.title = "SDK - Native"
        
        guard let nibObjects = Bundle.main.loadNibNamed("NativeAdView", owner: nil, options: nil),
              let adView = nibObjects.first as? VponNativeAdView else {
            fatalError("Could not load nib file for nativeAdView")
        }
        
        nativeAdView = adView
        adContainer.addSubview(adView)
        nativeAdView.translatesAutoresizingMaskIntoConstraints = false
        
        NSLayoutConstraint.activate([
            nativeAdView.widthAnchor.constraint(equalTo: adContainer.widthAnchor),
            nativeAdView.heightAnchor.constraint(equalTo: adContainer.heightAnchor)
        ])
        
        requestButtonDidTouch(requestButton)
    }
    
    // MARK: - Create VponAdRequest
    
    func createRequest() -> VponAdRequest {
        let request = VponAdRequest()
        request.setUserInfoGender(.male)                                                   // 性別
        request.setUserInfoBirthday(year: 2000, month: 08, day: 17)                        // 生日
        request.setContentUrl("https://www.google.com.tw/")                                // 內容
        request.setContentData(["key1": 1, "key2": true, "key3": "name", "key4": 123.31])  // 內容鍵值
        request.addFriendlyObstruction(UIView(), purpose: .mediaControls, description: "")
        
        VponAdRequestConfiguration.shared.testDeviceIdentifiers = ([ASIdentifierManager.shared().advertisingIdentifier.uuidString]) // 取得測試廣告
        VponAdRequestConfiguration.shared.tagForChildDirectedTreatment = .notForChildDirectedTreatment // 是否專為兒童投放
        VponAdRequestConfiguration.shared.tagForUnderAgeOfConsent = .notForUnderAgeOfConsent // 是否專為特定年齡投放
        VponAdRequestConfiguration.shared.maxAdContentRating = .general // 最高可投放的年齡(分類)限制
        
        return request
    }
    
    // MARK: Button Method
    
    @IBAction func requestButtonDidTouch(_ sender: UIButton) {
        sender.isEnabled = false
        
        adLoader = VponNativeAdLoader(licenseKey: "", rootViewController: self)
        adLoader?.delegate = self
        adLoader?.load(createRequest())
    }
}

extension VponSdkNativeViewController: VponNativeAdLoaderDelegate {
    func adLoader(_ adLoader: VponNativeAdLoader, didReceive nativeAd: VponNativeAd) {
        self.requestButton.isEnabled = true
        
        nativeAd.delegate = self
        
        (nativeAdView.headlineView as? UILabel)?.text = nativeAd.headline
        (nativeAdView.bodyView as? UILabel)?.text = nativeAd.body
        (nativeAdView.callToActionView as? UIButton)?.setTitle(nativeAd.callToAction, for: .normal)
        (nativeAdView.iconView as? UIImageView)?.image = nativeAd.icon?.image
        nativeAdView.callToActionView?.isUserInteractionEnabled = false
        nativeAdView.mediaView?.mediaContent = nativeAd.mediaContent
        
        if nativeAd.mediaContent?.hasVideoContent ?? false {
            nativeAd.mediaContent?.videoController?.delegate = self
        }
        
        // Necessary to show media content and make it clickable!
        nativeAdView.nativeAd = nativeAd
    }
    
    func adLoader(_ adLoader: VponNativeAdLoader, didFailToReceiveAdWithError error: Error) {
        self.requestButton.isEnabled = true
        print("adLoader didFailToReceiveAdWithError \(error.localizedDescription)")
    }
}

extension VponSdkNativeViewController: VponNativeAdDelegate {
    func nativeAdDidRecordImpression(_ nativeAd: VponNativeAd) {
        print("nativeAdDidRecordImpression")
    }
    
    func nativeAdDidRecordClick(_ nativeAd: VponNativeAd) {
        print("nativeAdDidRecordClick")
    }
}

extension VponSdkNativeViewController: VponVideoControllerDelegate {
    func videoControllerDidPlayVideo(_ videoController: VponVideoController) {
        print("videoControllerDidPlayVideo")
    }
    
    func videoControllerDidPauseVideo(_ videoController: VponVideoController) {
        print("videoControllerDidPauseVideo")
    }
    
    func videoControllerDidEndVideoPlayback(_ videoController: VponVideoController) {
        print("videoControllerDidEndVideoPlayback")
    }
    
    func videoControllerDidMuteVideo(_ videoController: VponVideoController) {
        print("videoControllerDidMuteVideo")
    }
    
    func videoControllerDidUnmuteVideo(_ videoController: VponVideoController) {
        print("videoControllerDidUnmuteVideo")
    }
}
